Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "ComponentReader.h"
+
+#include "Net/DataReplication.h"
+#include "Net/RepLayout.h"
+
+#include "EngineClasses/SpatialNetBitReader.h"
+#include "Interop/SpatialConditionMapFilter.h"
+#include "SpatialConstants.h"
+#include "Utils/SchemaUtils.h"
+#include "Utils/RepLayoutUtils.h"
+
+ComponentReader::ComponentReader(USpatialNetDriver* InNetDriver, FObjectReferencesMap& InObjectReferencesMap, TSet<UnrealObjectRef>& InUnresolvedRefs)
+	: PackageMap(InNetDriver->PackageMap)
+	, NetDriver(InNetDriver)
+	, ObjectReferencesMap(InObjectReferencesMap)
+	, UnresolvedRefs(InUnresolvedRefs)
+{
+}
+
+void ComponentReader::ApplyComponentData(const Worker_ComponentData& ComponentData, UObject* Object, USpatialActorChannel* Channel)
+{
+	Schema_Object* ComponentObject = Schema_GetComponentDataFields(ComponentData.schema_type);
+
+	ApplySchemaObject(ComponentObject, Object, Channel, true);
+}
+
+void ComponentReader::ApplyComponentUpdate(const Worker_ComponentUpdate& ComponentUpdate, UObject* Object, USpatialActorChannel* Channel)
+{
+	Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(ComponentUpdate.schema_type);
+
+	TArray<Schema_FieldId> ClearedIds;
+	ClearedIds.SetNum(Schema_GetComponentUpdateClearedFieldCount(ComponentUpdate.schema_type));
+	Schema_GetComponentUpdateClearedFieldList(ComponentUpdate.schema_type, ClearedIds.GetData());
+
+	ApplySchemaObject(ComponentObject, Object, Channel, false, &ClearedIds);
+}
+
+void ComponentReader::ApplySchemaObject(Schema_Object* ComponentObject, UObject* Object, USpatialActorChannel* Channel, bool bIsInitialData, TArray<Schema_FieldId>* ClearedIds)
+{
+	bool bAutonomousProxy = Channel->IsClientAutonomousProxy();
+
+	TArray<std::uint32_t> UpdateFields;
+	UpdateFields.SetNum(Schema_GetUniqueFieldIdCount(ComponentObject));
+	Schema_GetUniqueFieldIds(ComponentObject, UpdateFields.GetData());
+
+	if (UpdateFields.Num() == 0)
+	{
+		return;
+	}
+
+	FObjectReplicator& Replicator = Channel->PreReceiveSpatialUpdate(Object);
+
+	FRepState* RepState = Replicator.RepState;
+	TArray<FRepLayoutCmd>& Cmds = Replicator.RepLayout->Cmds;
+	TArray<FHandleToCmdIndex>& BaseHandleToCmdIndex = Replicator.RepLayout->BaseHandleToCmdIndex;
+	TArray<FRepParentCmd>& Parents = Replicator.RepLayout->Parents;
+
+	bool bIsServer = NetDriver->IsServer();
+	FSpatialConditionMapFilter ConditionMap(Channel, bAutonomousProxy);
+
+	TArray<UProperty*> RepNotifies;
+
+	for (std::uint32_t FieldId : UpdateFields)
+	{
+		// FieldId is the same as rep handle
```

Comment:
Is it? Is that always going to be the case?

Reply:
Yes that is always going to be the case. The schema is constructed in such a way that the handles always match their fields.